README.md
# Util library for MAX7219-powered LED matrix displays written in Rust for Linux (Raspberry Pi).

This is a `no_std` utility library on top of `max7219`-crate that allows you to easily display
text on dot matrix displays. **The main purpose of this lib is educational. There aren't mappings for all chars yet!**
Feel free to contribute on [Github](https://github.com/phip1611/max-7219-led-matrix-util)!

## `no_std` support vs usage on Raspberry Pi
By default, this crate requires `std` and provide easy setup functions using `gpio_cdev`-crate on Raspberry Pi for example.
If you need `no_std`, disable the default features.
#### Cargo.toml
```toml
[dependencies]
max-7219-led-matrix-util = "<latest-version>"
# or if you need `no_std`
max-7219-led-matrix-util = { version = "<latest-version>", default-features = false }
```

![demo](demo.gif)

## Usage example (`std`)
```rust
use max_7219_led_matrix_util::setup_adapter;
use max_7219_led_matrix_util::{prepare_display, show_moving_text_in_loop};

const NUM_DISPLAYS: usize = 4;

fn main() {
    println!("Demo for the 4-display device by AzDelivery. This is the device in the gif in the README.md.");
    println!();
    println!(
        "Provide 3 pins (gpio pin nums) please and connect all to the device: <data> <cs> <clk>"
    );
    println!("for example: '12 16 21'");
    println!();

    let args: Vec<String> = std::env::args().collect();
    assert_eq!(args.len(), 4, "Provide three args!");

    let data_pin = args[1].parse::<u32>().unwrap();
    let cs_pin = args[2].parse::<u32>().unwrap();
    let clk_pin = args[3].parse::<u32>().unwrap();

    println!("data={}, cs={}, clk={}", data_pin, cs_pin, clk_pin);

    // display adapter (std-feature, doesn't work in no_std)
    let mut display = setup_adapter("/dev/gpiochip0", NUM_DISPLAYS, data_pin, cs_pin, clk_pin);
    prepare_display(&mut display, NUM_DISPLAYS, 0x0F);
    show_moving_text_in_loop(
        &mut display,
        "HELLO 01 ABCDEF MAPA   ",
        NUM_DISPLAYS,
        // ms for each animation step
        50,
        // max_gap_width
        2
    );
}
```

### MSRV
The MSRV is `1.56.1`.

### Thanks to
Thanks to the creators of the `max7219`-crate!


CHANGELOG.md
# CHANGELOG

## v0.2.4 (2022-10-02)
- dependency updates
- small code improvements
- MSRV is `1.56.1`


src/lib.rs
//! Provides utility functions on top of "max7219"-crate to display data (like text) on a
//! MAX7219 powered matrix display.

#![deny(
    clippy::all,
    clippy::cargo,
    clippy::nursery,
    // clippy::restriction,
    // clippy::pedantic
)]
// now allow a few rules which are denied by the above statement
// --> they are ridiculous and not necessary
#![allow(
    clippy::fallible_impl_from,
    clippy::needless_doctest_main,
    clippy::redundant_pub_crate,
    clippy::suboptimal_flops
)]
#![deny(missing_docs)]
#![deny(missing_debug_implementations)]
#![deny(rustdoc::all)]
#![no_std]

#[cfg(feature = "std")]
extern crate std;

#[macro_use]
extern crate alloc;
use alloc::vec::Vec;

#[cfg(feature = "std")]
use crate::setup::Max7219;

#[cfg(feature = "std")]
use std::{thread::sleep, time::Duration};

#[cfg(feature = "std")]
use crate::encoding::encode_string;
use crate::mappings::SingleDisplayData;
#[cfg(feature = "std")]
use max7219::DecodeMode;

/// We use 8x8 square matrices (per single display)
pub const LED_SQUARE_MATRIX_DIM: usize = 8;

/// Maximum supported chained displays by MAX7219.
pub const MAX_DISPLAYS: usize = 16;

pub mod encoding;
pub mod mappings;
#[cfg(feature = "std")]
mod setup;
#[cfg(feature = "std")]
pub use setup::{setup as setup_adapter, Max7219 as Max7219Adapter};

/// Shift all row bits one to the left (to the next col). This way you can animate a moving text.
///
/// * `moving_bits` Vector with the data of all content to display. Each index describes
///                 the 8x8 bit data for a single display.
/// * `repeat` shift 1 bits on the very left to the ending of the vector. Without repeat
//            the vector will be all zeros after enough iterations.
pub fn shift_all_rows_one_bit_left(moving_bits: &mut [SingleDisplayData] /*, repeat: bool*/) {
    // move all bits to next position

    // so we iterate through the whole vector
    // note that probably only [0]..[DISPLAY_COUNT] are shown; this are the active displays
    // while that bits are shifted though the vec!

    let len = moving_bits.len();
    for display_i in 0..len {
        for row_i in 0..8 {
            // we need to shift to next segment if MSB per row is 1
            if moving_bits[display_i][row_i] & 0b10000000 != 0 {
                if display_i == 0
                /*&& repeat*/
                {
                    // to the last display
                    moving_bits[len - 1][row_i] |= 1;
                } else {
                    // to display from previous iteration
                    moving_bits[display_i - 1][row_i] |= 1;
                }
            }
            // shift all in row on to the left
            moving_bits[display_i][row_i] <<= 1;
        }
    }
}

/// Convenient function that turns on the display, clears the display
/// and sets the brightness to the highest possible value. It also sets
/// the DecodeMode to NoDecode which is necessary for displaying content on
/// the 8x8 matrix display. (Max7219 can also be used for 7 segment displays).
///
/// * `display` - mutable reference to Max7219 display driver
/// * `display_count` - count of displays connected to the MAX7219
/// * `intensity` - brightness for the display; value between `0x00` and `0x0F`
#[cfg(feature = "std")]
pub fn prepare_display(display: &mut Max7219, display_count: usize, intensity: u8) {
    let display_count = display_count % MAX_DISPLAYS;

    display.power_on().unwrap();
    for i in 0..display_count {
        display.set_decode_mode(i, DecodeMode::NoDecode).unwrap();
        display.clear_display(i).unwrap();
        display.set_intensity(i, intensity).unwrap();
    }
}

/// Shows a moving text in loop. After each iteration all bits are shifted one col to the left.
/// **Make sure to call `prepare_display()` first!**
///
/// * `display` - mutable reference to Max7219 display driver
/// * `text` - the text to display
/// * `display_count` - count of displays connected to the MAX7219
/// * `ms_sleep` - timeout after each iteration
/// * `gap_width` - (optional) set's the maximum width/count of empty cols between characters. 0 to deactivate.
///                 Downside is that spaces will be removed.
#[cfg(feature = "std")]
pub fn show_moving_text_in_loop(
    display: &mut Max7219,
    text: &str,
    display_count: usize,
    ms_sleep: u64,
    gap_width: Option<usize>,
) {
    let display_count = display_count % MAX_DISPLAYS;

    let raw_bits = encode_string(text);
    let mut display_data_vec: Vec<SingleDisplayData> = if let Some(gap_width) = gap_width {
        remove_gaps_in_display_text(&raw_bits, gap_width)
    } else {
        raw_bits
    };

    loop {
        for (i, display_data) in display_data_vec.iter().enumerate().take(display_count) {
            display.write_raw(i, display_data).unwrap();
        }
        sleep(Duration::from_millis(ms_sleep));
        // shift all rows one bit to the left
        shift_all_rows_one_bit_left(&mut display_data_vec);
    }
}

/// Iterates through the data and removes all gaps between symbols. A gap is two or more cols
/// after each other that are all zero. This way, text looks more natural, as letters are closer
/// together. Hence, we do not have a monospace font here. But (currently), also spaces are removed.
///
/// # Parameters
/// - `display_data_arr`: Processed Display Data with letters
/// - `min_gap_size`: Minimum gap size to preserve between symbols
pub fn remove_gaps_in_display_text(
    display_data_arr: &[SingleDisplayData],
    min_gap_size: usize,
) -> Vec<SingleDisplayData> {
    // all data in one single vector; each SingleDisplayData is transposed at first
    // => the vector contains all pixels but in (col1, colX, colN)-order.
    let display_data_expanded: Vec<u8> = display_data_arr
        .iter()
        // transpose: rows become cols
        .map(transpose_single_display_data)
        .flat_map(|display_data| display_data.into_iter())
        .collect();

    let preserve_at_begin = display_data_expanded
        .iter()
        .take_while(|col| **col == 0)
        .count();
    let preserve_at_end = display_data_expanded
        .iter()
        .rev()
        .take_while(|col| **col == 0)
        .count();

    // This variable holds the whole display data (col1, colX, ..., colN) but with
    // removed gaps in between. We keep empty cols at the begin.
    let mut shrinked_display_data_expanded = vec![0_u8; preserve_at_begin];

    let mut count_since_last_not_empty = 0;
    let skip_begin = preserve_at_begin;
    let skip_end = preserve_at_end;
    let take_elements = display_data_expanded.len() - skip_end - skip_begin + 1;

    // iterate and find gaps in between that can be removed
    for col in display_data_expanded.iter().take(take_elements).copied() {
        if col == 0 {
            count_since_last_not_empty += 1;
        } else {
            count_since_last_not_empty = 0;
        }

        // we preserve `min_gap_size` empty cols, after that, we skip them
        if count_since_last_not_empty <= min_gap_size {
            shrinked_display_data_expanded.push(col);
        }
    }

    // keep empty cols at end
    shrinked_display_data_expanded.extend_from_slice(&vec![0; preserve_at_end]);

    // now transform again to Vec<SingleDisplayData>
    // 1) check if length is multiple of eight
    let add_to_multiple_of_8 = 8 - (shrinked_display_data_expanded.len() % 8);
    shrinked_display_data_expanded.extend_from_slice(&vec![0; add_to_multiple_of_8]);

    // display data without gaps in-between ready for display.
    let shrinked_display_data = shrinked_display_data_expanded
        .as_slice()
        .chunks_exact(8)
        .map(|cols_8| {
            // transform to u8 array of length 8
            let mut transposed_display: SingleDisplayData = [0; 8];
            cols_8
                .iter()
                .copied()
                .enumerate()
                .for_each(|(i, col)| transposed_display[i] = col);
            transpose_single_display_data(&transposed_display)
        })
        .collect();

    shrinked_display_data
}

/// This does a transpose operation on the [`SingleDisplayData`]-Matrix and is a helper function for
/// [`remove_gaps_in_display_text`]. Cols become rows and rows become cols.
/// Example:
/// ```
/// let _ = [
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
///     0b10000000,
/// ];
/// ```
/// becomes
/// ```
/// let _ = [
/// 0b11111111,
/// 0,
/// 0,
/// 0,
/// 0,
/// 0,
/// 0,
/// 0,
/// ];
/// ```
#[allow(clippy::needless_range_loop)]
pub fn transpose_single_display_data(data: &SingleDisplayData) -> SingleDisplayData {
    let mut transposed_data: SingleDisplayData = [0; 8];
    for col_i in 0..8 {
        // the data/bits of the current col
        let mut col = 0;
        for row_i in 0..8 {
            // we get the current col value (bit) at index "data[row_i][col_i]"
            // bit by bit. We move the current bit to the lowest index via
            // bit shifting and do a bitwise and with 1.
            let col_bit = (data[row_i] >> (7 - col_i)) & 1;
            // now we first shift all col bits from previous iterations one index
            // to the left and then we add our current bit to the col at the lowest index.
            col = (col << 1) | col_bit;
        }
        transposed_data[col_i] = col;
    }
    transposed_data
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_move_all_bits_one_col_left() {
        let data_dis_0 = [
            0b01000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000,
            0b00000000,
        ];
        let data_dis_1 = [
            0b11000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000, 0b00000000,
            0b00000000,
        ];
        let mut data = Vec::new();
        data.push(data_dis_0);
        data.push(data_dis_1);

        shift_all_rows_one_bit_left(&mut data);

        let first_row_dis_0_expected = 0b10000001;
        let first_row_dis_1_expected = 0b10000000;
        let first_row_dis_0_actual = data[0][0];
        let first_row_dis_1_actual = data[1][0];

        assert_eq!(first_row_dis_0_actual, first_row_dis_0_expected);
        assert_eq!(first_row_dis_1_actual, first_row_dis_1_expected);
    }

    #[test]
    fn test_transpose_single_display_data() {
        let input = [
            0b0100_0000,
            0b0100_0000,
            0b0100_0000,
            0b0100_0000,
            0b0000_0011,
            0b0000_0011,
            0b0000_0011,
            0b0000_0011,
        ];
        let expected = [
            0b0000_0000,
            0b1111_0000,
            0b0000_0000,
            0b0000_0000,
            0b0000_0000,
            0b0000_0000,
            0b0000_1111,
            0b0000_1111,
        ];

        let actual = transpose_single_display_data(&input);

        for i in 0..input.len() {
            assert_eq!(
                actual[i], expected[i],
                "swap_cols_to_rows() doesn't transposed the matrix properly at index {}! is={:#b}, expected={:#b}",
                i, actual[i], expected[i]
            );
        }
    }

    #[test]
    fn test_remove_gaps_in_display_text() {
        let vec = vec![
            [
                0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000,
                0b10000000,
            ],
            [
                0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000,
                0b10000000,
            ],
        ];
        let expected = vec![
            [
                0b10010000, 0b10010000, 0b10010000, 0b10010000, 0b10010000, 0b10010000, 0b10010000,
                0b10010000,
            ],
            [
                // TODO remove last if only empty?!
                0, 0, 0, 0, 0, 0, 0, 0,
            ],
        ];
        let actual = remove_gaps_in_display_text(&vec, 2);
        for i in 0..2 {
            for j in 0..8 {
                assert_eq!(
                    actual[i][j], expected[i][j],
                    "expected: {:#b}, is: {:#b}",
                    expected[i][j], vec[i][j]
                );
            }
        }
    }
}


src/setup.rs
use embedded_hal::digital::v2::OutputPin;
use gpio_cdev::{Chip, LineHandle, LineRequestFlags};
use max7219::connectors::PinConnector;
use max7219::MAX7219;

/// The type of the driver.
pub type Max7219 = MAX7219<PinConnector<LHandle, LHandle, LHandle>>;

// setup glue structure between crates "embedded_hal", "gpio_cdev" and "max7219 "
/// Wrapper around [`gpio_cdev::LineHandle`] that implements [`embedded_hal::digital::v2::OutputPin`].
#[derive(Debug)]
pub struct LHandle(LineHandle);
impl OutputPin for LHandle {
    type Error = ();

    fn set_low(&mut self) -> Result<(), Self::Error> {
        self.0.set_value(0).map_err(|_| ())
    }

    fn set_high(&mut self) -> Result<(), Self::Error> {
        self.0.set_value(1).map_err(|_| ())
    }
}
// setup glue structure done

/// Set's up the MAX7219-driver using "gpio_cdev". This only works on Linux. It works for example
/// on Raspberry. The communication uses SPI protocol. This does not require the "SPI Module" on
/// Raspberry Pi to be activated. This works via regular GPIO pins.
///
/// * `gpio_dev` the gpio device. Probably "/dev/gpiochip0"
/// * `num_displays` number of displays
/// * `data_pin` number of GPIO pin used as data pin
/// * `clk_pin` number of GPIO pin used as clock pin
/// * `cs_pin` number of GPIO pin used as cs (chip select) pin
pub fn setup(
    gpio_dev: &str,
    num_displays: usize,
    data_pin: u32,
    cs_pin: u32,
    clk_pin: u32,
) -> Max7219 {
    let mut gpio = Chip::new(gpio_dev).unwrap();
    let data_pin = gpio
        .get_line(data_pin)
        .unwrap()
        .request(LineRequestFlags::OUTPUT, 0, "spi-data-pin")
        .unwrap();
    let cs_pin = gpio
        .get_line(cs_pin)
        .unwrap()
        .request(LineRequestFlags::OUTPUT, 0, "spi-cs-pin")
        .unwrap();
    let clk_pin = gpio
        .get_line(clk_pin)
        .unwrap()
        .request(LineRequestFlags::OUTPUT, 0, "spi-clk-pin")
        .unwrap();

    let data_pin = LHandle(data_pin);
    let clk_pin = LHandle(clk_pin);
    let cs_pin = LHandle(cs_pin);

    MAX7219::from_pins(num_displays, data_pin, cs_pin, clk_pin).unwrap()
}


src/encoding.rs
//! Encoding utility functions that helps you to print symbols easily on a
//! MAX7219-powered LED matrix.

use crate::mappings::*;
use alloc::vec::Vec;

/// Encodes a char to its bit-representation on a single display. This means a 8x8 bit matrix.
/// Currently only a very limited alphabet is available and only capital letters. Unknown chars
/// are mapped to SPACE (empty).
#[allow(clippy::wildcard_in_or_patterns)]
pub const fn encode_char(c: char) -> SingleDisplayData {
    // currently we only support cap chars
    let c = c.to_ascii_uppercase();
    match c {
        'A' => CAP_A,
        'B' => CAP_B,
        'C' => CAP_C,
        'D' => CAP_D,
        'E' => CAP_E,
        'F' => CAP_F,
        'G' => CAP_G,
        'H' => CAP_H,
        'I' => CAP_I,
        'J' => CAP_J,
        'K' => CAP_K,
        'L' => CAP_L,
        'M' => CAP_M,
        'N' => CAP_N,
        'O' => CAP_O,
        'P' => CAP_P,
        'Q' => CAP_Q,
        'R' => CAP_R,
        'S' => CAP_S,
        'T' => CAP_T,
        'U' => CAP_U,
        'V' => CAP_V,
        'W' => CAP_W,
        'X' => CAP_X,
        'Y' => CAP_Y,
        'Z' => CAP_Z,

        '0' => ZERO,
        '1' => ONE,

        '.' => DOT,
        '!' => EXCLAMATION_MARK,

        ' ' | _ => SPACE,
    }
}

/// Encodes each char of a string to its bit-representation. The resulting vector should be understood
/// as the data to be displayed on a display chain that is as long as the coupled/chained displays
/// that are powered by the MAX7219.
/// So if you have 4 displays and 10 chars then you could only display indices zero to three and
/// bit shift all rows one to the left per iteration. This way you get a smooth transition/animation.
/// Don't forget about the sleep-timeout per iteration!
pub fn encode_string(s: &str) -> Vec<SingleDisplayData> {
    s.chars().into_iter().map(encode_char).collect()
}


src/mappings.rs
//! Provides mappings for a mono spaced font I created myself.
//! The font is not complete and many symbols are missing.
//! This ist just an demonstration how it could be done.
//!
//! Each symbol is exactly `SingleDisplayData * SingleDisplayData` big.

use crate::LED_SQUARE_MATRIX_DIM;

/// We have 8 rows and 8 bits per row.
pub type SingleDisplayData = [u8; LED_SQUARE_MATRIX_DIM];

/// Capital letter A mapping.
pub const CAP_A: SingleDisplayData = [
    0b00111000, 0b01000100, 0b01000100, 0b01000100, 0b01111100, 0b01000100, 0b01000100, 0b01000100,
];
/// Capital letter B mapping.
pub const CAP_B: SingleDisplayData = [
    0b01111000, 0b01000100, 0b01000100, 0b01111000, 0b01000100, 0b01000100, 0b01000100, 0b01111000,
];
/// Capital letter C mapping.
pub const CAP_C: SingleDisplayData = [
    0b01111100, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01111100,
];
/// Capital letter D mapping.
pub const CAP_D: SingleDisplayData = [
    0b01111000, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b01111000,
];
/// Capital letter E mapping.
pub const CAP_E: SingleDisplayData = [
    0b01111100, 0b01000000, 0b01000000, 0b01111100, 0b01000000, 0b01000000, 0b01000000, 0b01111100,
];
/// Capital letter F mapping.
pub const CAP_F: SingleDisplayData = [
    0b01111100, 0b01000000, 0b01000000, 0b01111100, 0b01000000, 0b01000000, 0b01000000, 0b01000000,
];
/// Capital letter G mapping.
pub const CAP_G: SingleDisplayData = [
    0b01111000, 0b11000100, 0b10000100, 0b10000000, 0b10011100, 0b10000100, 0b11000100, 0b01111100,
];
/// Capital letter H mapping.
pub const CAP_H: SingleDisplayData = [
    0b01000100, 0b01000100, 0b01000100, 0b01111100, 0b01000100, 0b01000100, 0b01000100, 0b01000100,
];
/// Capital letter I mapping.
pub const CAP_I: SingleDisplayData = [
    0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000,
];
/// Capital letter J mapping.
pub const CAP_J: SingleDisplayData = [0; 8]; // TODO
/// Capital letter K mapping.
pub const CAP_K: SingleDisplayData = [
    0b01000100, 0b01001000, 0b01010000, 0b01100000, 0b01010000, 0b01001000, 0b01000100, 0b01000010,
];
/// Capital letter L mapping.
pub const CAP_L: SingleDisplayData = [
    0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01000000, 0b01111100,
];
/// Capital letter M mapping.
pub const CAP_M: SingleDisplayData = [
    0b10000010, 0b11000110, 0b10101010, 0b10111010, 0b10010010, 0b10000010, 0b10000010, 0b10000010,
];
/// Capital letter N mapping.
pub const CAP_N: SingleDisplayData = [
    0b01000100, 0b01100100, 0b01110100, 0b01010100, 0b01011100, 0b01001100, 0b01001100, 0b01000100,
];
/// Capital letter O mapping.
pub const CAP_O: SingleDisplayData = [
    0b00011000, 0b00100100, 0b01000010, 0b01000010, 0b01000010, 0b01000010, 0b00100100, 0b00011000,
];
/// Capital letter P mapping.
pub const CAP_P: SingleDisplayData = [
    0b01111000, 0b01000100, 0b01000100, 0b01000100, 0b01111000, 0b01000000, 0b01000000, 0b01000000,
];
/// Capital letter Q mapping.
pub const CAP_Q: SingleDisplayData = [0; 8]; // TODO
/// Capital letter R mapping.
pub const CAP_R: SingleDisplayData = [
    0b01111000, 0b01000100, 0b01000100, 0b01111000, 0b01100000, 0b01010000, 0b01001000, 0b01000100,
];
/// Capital letter S mapping.
pub const CAP_S: SingleDisplayData = [
    0b00011100, 0b00100000, 0b01000000, 0b00110000, 0b00001000, 0b00000100, 0b00000100, 0b01111000,
];
/// Capital letter T mapping.
pub const CAP_T: SingleDisplayData = [
    0b11111110, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000,
];
/// Capital letter U mapping.
pub const CAP_U: SingleDisplayData = [
    0b01000010, 0b01000010, 0b01000010, 0b01000010, 0b01000010, 0b01000010, 0b01000010, 0b00111100,
];
/// Capital letter V mapping.
pub const CAP_V: SingleDisplayData = [0; 8]; // TODO
/// Capital letter W mapping.
pub const CAP_W: SingleDisplayData = [0; 8]; // TODO
/// Capital letter X mapping.
pub const CAP_X: SingleDisplayData = [0; 8]; // TODO
/// Capital letter Y mapping.
pub const CAP_Y: SingleDisplayData = [0; 8]; // TODO
/// Capital letter Z mapping.
pub const CAP_Z: SingleDisplayData = [
    0b01111110, 0b00000010, 0b00000100, 0b00001000, 0b00010000, 0b00100000, 0b01000000, 0b01111110,
];
/// Number 0 mapping.
pub const ZERO: SingleDisplayData = [
    0b00111000, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b01000100, 0b00111000,
];
/// Number 1 mapping.
pub const ONE: SingleDisplayData = [
    0b00000100, 0b00011100, 0b00000100, 0b00000100, 0b00000100, 0b00000100, 0b00000100, 0b00000100,
];
/// " " character mapping.
pub const SPACE: SingleDisplayData = [0; 8];
/// "." character mapping.
pub const DOT: SingleDisplayData = [0, 0, 0, 0, 0, 0, 0, 0b00010000];
/// "!" character mapping.
pub const EXCLAMATION_MARK: SingleDisplayData = [
    0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0b00010000, 0, 0b00010000,
];


examples/demo.rs
//! Demo for the 4-display device by AzDelivery. This is the device in the gif in the README.md.
//! (https://www.az-delivery.de/products/4-x-64er-led-matrix-display).

use max_7219_led_matrix_util::setup_adapter;
use max_7219_led_matrix_util::{prepare_display, show_moving_text_in_loop};

const NUM_DISPLAYS: usize = 4;

fn main() {
    println!("Demo for the 4-display device by AzDelivery. This is the device in the gif in the README.md.");
    println!();
    println!(
        "Provide 3 pins (gpio pin nums) please and connect all to the device: <data> <cs> <clk>"
    );
    println!("for example: '12 16 21'");
    println!();

    let args: Vec<String> = std::env::args().collect();
    assert_eq!(args.len(), 4, "Provide three args!");

    let data_pin = args[1].parse::<u32>().unwrap();
    let cs_pin = args[2].parse::<u32>().unwrap();
    let clk_pin = args[3].parse::<u32>().unwrap();

    println!("data={}, cs={}, clk={}", data_pin, cs_pin, clk_pin);

    // display adapter (std-feature, doesn't work in no_std)
    let mut display = setup_adapter("/dev/gpiochip0", NUM_DISPLAYS, data_pin, cs_pin, clk_pin);
    prepare_display(&mut display, NUM_DISPLAYS, 0x0F);
    show_moving_text_in_loop(
        &mut display,
        "HALLO LEON!!!  ",
        NUM_DISPLAYS,
        // ms for each animation step
        50,
        // gap_width
        None,
    );
}
